### PR TITLE
React 19

### LINKED ISSUE #2756: compat with react 19
React pdf is not compatible with react 19, will you make the library compat with this version ?

### PR DESCRIPTION
**Pull request by alexandernanberg**
Fixes #2756

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -46,7 +46,6 @@
     "@typescript-eslint/parser": "^6.21.0",
     "@vitejs/plugin-react": "^4.2.1",
     "babel-plugin-add-module-exports": "^1.0.0",
-    "canvas": "^2.11.2",
```
Comment (alexandernanberg) [inline package.json:49]: Completely unable to install this dep, we'll revisit this later

File: packages/renderer/rollup.config.js [hunk 2]
```diff
@@ -39,20 +38,23 @@ const getExternal = ({ browser }) => [
...
+  // ),
   ...Object.keys(pkg.peerDependencies),
 ];
 
 const getPlugins = ({ browser, declarationDests, minify = false }) => [
   json(),
   ...(browser ? [ignore(['fs', 'path', 'url'])] : []),
-  alias({
-    entries: {
-      'react-reconciler': 'react-reconciler/cjs/react-reconciler.production.js',
```
Comment (alexandernanberg) [inline packages/renderer/rollup.config.js]: Not really sure why this was needed, not bundling `react-reconciler` seems like a better way forward?

**Comment by alexandernanberg** [comment]
Currently both `react-reconciler.development.js` and `react-reconciler.production.js` are included in the bundle, which isn't optimal. I don't see an alternative though, since `jsx-runtime` in dev requires the dev reconciler

**Comment by joacub** [comment]
no solutions ?

**Comment by diegomura** [comment]
This is a solution for react19 at the expense of dropping support for all other versions for what I understand. Is that correct?

**Comment by alexandernanberg** [comment]
Yes, that is my understanding. That said, I've not tested this branch with React 18

**Comment by Ortieez** [comment]
Hi, is there an estimate of when this PR will be merged into the main branch? Our project kind of depends on this. Thanks in advance.

**Comment by diegomura** [comment]
@alexandernanberg [here's](https://github.com/diegomura/react-pdf/tree/dm/react-19) the branch where I'm pushing that. My idea is to ship both reconcilers, and dynamically use one or the other based on React version. Code there is the one included in this PR, but different structured because of that reason. However, when trying it out I get the following errors. 

Did you get those on your testing by any chance? Would you have the time to take a look at my branch? I'm planning to add you as co-author of it anyways :) Thanks in advance

Screenshot 2024-11-15 at 02 45 19

**Comment by alexandernanberg** [comment]
> is this updated to what's published in npm?

Yes, but cut a new `4.0.0-canary-5` just to be sure. Still works as expected in my app.

> Did you get those on your testing by any chance?

I don't remember exactly, but I did have some issues when trying to use the prod reconciler with dev react or vice versa, the prod bundles are slimmed down and don't include everything. The React packages use dynamic import/requires https://github.com/facebook/react/blob/main/packages/react-reconciler/npm/index.js 

I'll take a look at your branch when I have some time to spare!

**Comment by diegomura** [comment]
Will close this in favor of https://github.com/diegomura/react-pdf/pull/2941 where I kept support for all the previous React versions too. @alexandernanberg thanks so much for posting this. It was very helpful. Added you as a co-author of the PR although I don't know your email to correctly add the commit message

**Comment by alexandernanberg** [comment]
Cool, look forward to trying it out! Added my email to my GitHub profile but probably annoying to change it afterwards

**Pull request closed without merging**